drivechain-server/bdk/bdk.go

```diff
@@ -293,3 +300,163 @@ type Transaction struct {
 	Sent     btcutil.Amount `json:"sent"`
 	TXID     string         `json:"txid"`
 }
+
+// NewWallet either creates a new wallet or loads an existing one
+func NewWallet(datadir, network, electrum, passphrase string) (*Wallet, error) {
+	keyFile := filepath.Join(datadir, "wallet.key")
+	var xprv string
+
+	if _, err := os.Stat(keyFile); err == nil {
+		xprv, err = loadExistingWallet(keyFile, passphrase)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load existing wallet: %w", err)
+		}
+	} else {
+		xprv, err = createNewWallet(keyFile, passphrase)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create new wallet: %w", err)
+		}
+	}
+
+	return &Wallet{
+		Descriptor: fmt.Sprintf("wpkh(%s/84h/1h/0h/0/*)", xprv),
+		Network:    network,
+		Datadir:    datadir,
+		Electrum:   electrum,
+	}, nil
+}
+
+func loadExistingWallet(keyFile, passphrase string) (string, error) {
+	encryptedMnemonic, err := os.ReadFile(keyFile)
+	if err != nil {
+		return "", fmt.Errorf("failed to read existing key file: %w", err)
+	}
+
+	mnemonic, err := decryptKey(encryptedMnemonic, passphrase)
+	if err != nil {
+		return "", fmt.Errorf("failed to decrypt mnemonic: %w", err)
+	}
+
+	bdkCliPath, err := getBdkCliPath()
+	if err != nil {
+		return "", fmt.Errorf("failed to get bdk-cli path: %w", err)
+	}
+
+	res, err := exec.Command(bdkCliPath, "key", "restore", "--mnemonic", mnemonic).Output()
+	if err != nil {
+		return "", fmt.Errorf("failed to restore key: %w", err)
+	}
+
+	xprv := gjson.GetBytes(res, "xprv").String()
+	if xprv == "" {
+		return "", fmt.Errorf("failed to parse xprv from bdk-cli output")
+	}
+
+	return xprv, nil
+}
+
+func createNewWallet(keyFile, passphrase string) (string, error) {
+	bdkCliPath, err := getBdkCliPath()
+	if err != nil {
+		return "", fmt.Errorf("failed to get bdk-cli path: %w", err)
+	}
+
+	res, err := exec.Command(bdkCliPath, "key", "generate").Output()
+	if err != nil {
+		return "", fmt.Errorf("failed to generate new key: %w", err)
+	}
+
+	mnemonic := gjson.GetBytes(res, "mnemonic").String()
+	if mnemonic == "" {
+		return "", fmt.Errorf("failed to parse mnemonic from bdk-cli output")
+	}
+
+	xprv := gjson.GetBytes(res, "xprv").String()
+	if xprv == "" {
+		return "", fmt.Errorf("failed to parse xprv from bdk-cli output")
+	}
+
+	var dataToSave []byte
+	if passphrase != "" {
+		encryptedMnemonic, err := encryptKey(mnemonic, passphrase)
+		if err != nil {
+			return "", fmt.Errorf("failed to encrypt mnemonic: %w", err)
+		}
+		dataToSave = encryptedMnemonic
+	} else {
+		dataToSave = []byte(mnemonic)
+	}
+
+	err = os.WriteFile(keyFile, dataToSave, 0600)
+	if err != nil {
+		return "", fmt.Errorf("failed to save mnemonic to file: %w", err)
+	}
+
+	return xprv, nil
+}
+
+func encryptKey(key, passphrase string) ([]byte, error) {
```

Comment:
A few thoughts here. 

1. We should be documenting what we're doing. When dealing with encryption there's always so many things to think about. Key stretching algorithms, encryption algorithms, cipher modes, block sizes... And more!
2. I've always strived to avoid implementing cryptography stuff, because it's so daunting (and when dealing with real money, really sensitive) to get it right. How sure are you that these parameters are "right"? One alternative to look into is using [`age`](https://github.com/FiloSottile/age). It's a simple encryption protocol created by one of the Golang cryptography maintainers, that purposefully has no knobs to turn. It's available both as a CLI tool and a Golang library. Could be something to consider!

Reply:
Not saying that we _have_ to change how things are done here. Just want us to have a (short) conversation about it before we settle the details.

Reply:
You're right, it is a bit daunting and a lot of times it's hard to even argue why certain param values are better than others, etc. I'll add some comments for the least obvious stuff.

Age looks perfect for this use case, at least for the time being. I'll give it a try and if it's a good fit I'll switch implementations.